File: day10/3_sortMatrix_diagonally.py
# created by KUMAR SHANU

# 3. Sort the Matrix Diagonally
# https://leetcode.com/problems/sort-the-matrix-diagonally/

# Solution 1
"""
store each diagonal into the dictionary and sort.
"""
from collections import defaultdict


class Solution:
    def diagonalSort(self, mat: List[List[int]]) -> List[List[int]]:
        d = defaultdict(list)
        for i in range(len(mat)):
            for j in range(len(mat[0])):
                # store diagonal into dictionary
                diff = i - j
                d[diff].append(mat[i][j])

        for i in range(len(mat)):
            for j in range(len(mat[0])):
                # sort the diagonal
                diff = i - j
                d[diff].sort()
                # insert it back to matrix
                mat[i][j] = d[diff].pop(0)

        return mat


# Solution 2
"""
Using dict and heap
"""
import heapq


class Solution:
    def diagonalSort(self, mat: List[List[int]]) -> List[List[int]]:
        d = defaultdict(list)
        for i in range(len(mat)):
            for j in range(len(mat[0])):
                # push diagonal into heap
                diff = i - j
                heapq.heappush(d[diff], mat[i][j])
        for i in range(len(mat)):
            for j in range(len(mat[0])):
                diff = i - j
                mat[i][j] = heapq.heappop(d[diff])

        return mat

File: day0010/3-Sort-Matrix-Diagonally.java
/*
Auther - Saurabh Verma
Problem link - https://leetcode.com/problems/sort-the-matrix-diagonally/
Time Complexity- (m*n)log(min(m,n))
*/

class Solution {
    public int[][] diagonalSort(int[][] mat) {
        Map<Integer, PriorityQueue<Integer>> map = new HashMap<>();
        for(int i=0; i<mat.length; i++){
            for(int j=0; j<mat[i].length; j++){
               int diff = i - j;
                if(map.containsKey(diff)){
                    map.get(diff).add(mat[i][j]);
                }else{
                    map.put(diff, new PriorityQueue());
                    map.get(diff).add(mat[i][j]);
                }
            }
        }
        for (int i = 0; i < mat.length; i++) {
            for (int j = 0; j < mat[i].length; j++) {
                mat[i][j] = map.get(i-j).poll();
            }
           
        }
        return mat;
        
    }
}
